pkg/query/iter.go

```diff
@@ -587,3 +593,56 @@ func (it *dedupSeriesIterator) Err() error {
 	}
 	return it.b.Err()
 }
+
+// counterDedupAdjustSeriesIterator is used when we deduplicate counter.
+// It makes sure we always adjust for the latest seen last counter value for all replicas.
+// Let's consider following example:
+//
+// Replica 1 counter scrapes: 20    30    40    Nan      -     0     5
+// Replica 2 counter scrapes:    25    35    45     Nan     -     2
+//
+// Now for downsampling purposes we are accounting the resets so our replicas before going to dedup iterator looks like this:
+//
+// Replica 1 counter total: 20    30    40   -      -     40     45
+// Replica 2 counter total:    25    35    45    -     -     47
+//
+// Now if at any point we will switch our focus from replica 2 to replica 1 we will experience lower value than previous,
+// which will trigger false positive counter reset in PromQL.
+//
+// We mitigate this by taking always adjusting for the "behind" replica value to be not smaller than highest sample seen.
+// This is also what is closest to the truth (last seen counter value on this target).
+//
+// This short-term solution to mitigate https://github.com/thanos-io/thanos/issues/2401.
+// TODO(bwplotka): Find better deduplication algorithm that does not require knowledge if the given
+// series is counter or not: https://github.com/thanos-io/thanos/issues/2547.
+type counterDedupAdjustSeriesIterator struct {
+	storage.SeriesIterator
+
+	lastV  float64
+	adjust float64
+}
+
+func newCounterDedupAdjustSeriesIterator(iter storage.SeriesIterator) storage.SeriesIterator {
+	return &counterDedupAdjustSeriesIterator{
+		SeriesIterator: iter,
+		lastV:          -1 * math.MaxFloat64,
+	}
+
+}
+
+func (it *counterDedupAdjustSeriesIterator) Next() bool {
+	if it.SeriesIterator.Next() {
+		_, v := it.SeriesIterator.At()
+		if it.lastV > v {
```

Comment:
Hmm, I was thinking but maybe a stupid question: shouldn't we set `it.adjust` back to 0 if we will jump back to a replica with a higher counter value again at some point during the iteration? Wouldn't that lead to higher counter values than usual without such an adjustment?

Reply:
Hmm, you might be right. Adjust has to be cumulative, but per replica not globally, otherwise, we are adjusting true value... This might mean we need to

1. Actually hook inside the dedup iterator and adjust per replica (fine)
2, Or reset adjust as you propose (easier, but we lost the info that replica was behind).

Also if we do 1 it's a bit broken anyway as we know about WHEN & WHAT to adjust only if we are lucky that replica changes (: Still it's better than what we had before, plus better than any solution in the world for this (e.g dedup on ingest Cortex have)

cc @brancz @brian-brazil 

Will think about it a bit more, thanks for spotting this!

Reply:
I fixed this problem, but changed code a little bit, please review @GiedriusS and thanks for spotting this! I added test cases for this